modules/sirdataRtdProvider.js

```diff
@@ -0,0 +1,519 @@
+          site: {
+            ext: {
+              data: {
+                sd_rtd: categories,
+              }
+            }
+          },
+          user: {
+            ext: {
+              data: {
+                sd_rtd: segments
+              }
+            }
+          }
+        }
+      }
+    });
+  } catch (err) {
+    utils.logError(err.message)
+  }
+
+  return true;
+}
+
+export function setBidderFpd(bidder, segments, categories, globalConfig) {
+  try {
+    globalConfig.setBidderConfig({
+      bidders: [bidder],
+      config: {
+        fpd: {
+          context: {
+            data: {
+              sd_rtd: categories
+            }
+          },
+          user: {
+            data: {
+              sd_rtd: segments
+            }
+          }
+        }
+      }
+    });
+  } catch (err) {
+    utils.logError(err.message)
+  }
+  return true;
+}
+
+export function loadCustomFunction (todo, adUnit, list, data, bid) {
+  try {
+    if (typeof todo == 'function') {
+      todo(adUnit, list, data, bid);
+    }
+  } catch (er) { utils.logError(er); }
+  return true;
+}
+
+export function getSegAndCatsArray(data, minScore) {
+  var sirdataData = {'segments': [], 'categories': []};
+  minScore = minScore && typeof minScore == 'number' ? minScore : 30;
+  try {
+    if (data && data.contextual_categories && minScore) {
+      Object.entries(data.contextual_categories).forEach(([cat, value]) => {
+        if (value >= minScore && sirdataData.categories.indexOf(cat) === -1) {
+          sirdataData.categories.push(cat.toString());
+        }
+      });
+    }
+  } catch (e) {}
+  try {
+    if (data && data.segments) {
+      Object.entries(data.segments).forEach(([entry, segment]) => {
+        if (sirdataData.segments.indexOf(sirdataData.segment) === -1) {
+          sirdataData.segments.push(segment.toString());
+        }
+      });
+    }
+  } catch (e) {}
+  return sirdataData;
+}
+
+export function addSegmentData(adUnits, data, config, onDone, globalConfig) {
+  config = config || {};
+  config.params = config.params || {};
+  const minScore = config.params.hasOwnProperty('contextualMinRelevancyScore') ? config.params.contextualMinRelevancyScore : 30;
+  var sirdataData = getSegAndCatsArray(data, minScore);
+
+  if (!sirdataData || (sirdataData.segments.length < 1 && sirdataData.categories.length < 1)) { utils.logError('no cats'); onDone(); return adUnits; }
+
+  const sirdataList = sirdataData.segments.concat(sirdataData.categories);
+
+  var curationData = {'segments': [], 'categories': []};
+  var curationId = '1';
+  const biddersParamsExist = (!!(config.params && config.params.bidders));
+
+  // Global ortb2
+  if (!biddersParamsExist) {
+    setGlobalOrtb2(sirdataData.segments, sirdataData.categories, globalConfig);
+  }
+
+  // Google targeting
+  if (typeof window.googletag !== 'undefined' && (config.params.setGptKeyValues || !config.params.hasOwnProperty('setGptKeyValues'))) {
+    // For curation GG is pid 27449
```

Comment:
what is gg pid?

Reply:
Google partner id. The RTD module is to be used by a publisher and segments ids and category ids returned by the requested API are salted and hashed. Only the publisher knows what the ids means  and can use them with the bidder for targeting so we avoid data leakage. A few tech provider are also able to provide targeting accross publishers and suppliers (aka curators) and the publisher can request us to also deliver segment and catgory ids that the SSP/bidder will "understand". Default curatorId for Google is 27449 in our system